src/face_op.cpp

```diff
@@ -192,38 +247,120 @@ void Manifold::Impl::Face2Tri(const Vec<int>& faceEdge,
+                                    static_cast<uint64_t>(ref.faceID);
+                   ++vertDegree[halfedge_[edge].startVert];
+                 }
+               }
+             });
+
+  Vec<size_t> newHalf2Old(halfedge_.size());
+  sequence(newHalf2Old.begin(), newHalf2Old.end());
+  stable_sort(
+      newHalf2Old.begin(), newHalf2Old.end(),
+      [&edgeFace](size_t a, size_t b) { return edgeFace[a] < edgeFace[b]; });
+  newHalf2Old.resize(std::find_if(countAt(0_uz), countAt(halfedge_.size()),
+                                  [&](const size_t i) {
+                                    return edgeFace[newHalf2Old[i]] == remove;
+                                  }) -
+                     countAt(0_uz));
+
+  Vec<Halfedge> newHalfedge(newHalf2Old.size());
+  for_each_n(policy, countAt(0_uz), newHalf2Old.size(),
+             [&](size_t i) { newHalfedge[i] = halfedge_[newHalf2Old[i]]; });
+
+  Vec<int> faceEdge(1, 0);
+  for (size_t i = 1; i < newHalf2Old.size(); ++i) {
+    if (edgeFace[newHalf2Old[i]] != edgeFace[newHalf2Old[i - 1]]) {
+      faceEdge.push_back(i);
+    }
   }
+  const int numFace = faceEdge.size();
+  faceEdge.push_back(newHalf2Old.size());
 
-  PolygonsIdx polys;
-  int startEdge = 0;
-  int thisEdge = startEdge;
-  while (1) {
-    if (thisEdge == startEdge) {
-      if (vert_edge.empty()) break;
-      startEdge = vert_edge.begin()->second;
-      thisEdge = startEdge;
-      polys.push_back({});
+  if (numFace < 4) {
+    MarkFailure(Error::NoError);
+    return;  // empty
+  }
+
+  Vec<TriRef> halfedgeRef(halfedge_.size());
+  Vec<vec3> oldFaceNormal = std::move(faceNormal_);
+  faceNormal_.resize(numFace);
+
+  std::atomic<uint64_t> faceData(0);
+  Vec<int> faceEdge2(faceEdge.size());
+  faceEdge2[0] = 0;
+  Vec<Halfedge> newHalfedge2(newHalfedge.size());
+  for_each_n(policy, countAt(0_uz), numFace, [&](size_t inFace) {
+    std::vector<std::vector<int>> polys =
+        AssembleHalfedges(newHalfedge.cbegin() + faceEdge[inFace],
+                          newHalfedge.cbegin() + faceEdge[inFace + 1]);
+    Vec<Halfedge> polys2;
+    for (const auto& poly : polys) {
+      int start = -1;
+      int last = -1;
+      for (const int vert : poly) {
+        if (vertDegree[vert] < 3) continue;
+        if (start == -1) {
+          start = vert;
+        } else if (last != -1) {
+          polys2.push_back({last, vert});
+        }
+        last = vert;
+      }
+      polys2.push_back({last, start});
     }
-    int vert = (start + thisEdge)->startVert;
-    polys.back().push_back({projection * vertPos_[vert], vert});
-    const auto result = vert_edge.find((start + thisEdge)->endVert);
-    DEBUG_ASSERT(result != vert_edge.end(), topologyErr, "non-manifold edge");
-    thisEdge = result->second;
-    vert_edge.erase(result);
+
+    const int numEdge = polys2.size();
+    if (numEdge < 3) return;
+    const uint64_t inc = (1LL << 32) + numEdge;
```

Comment:
We should probably document this so that whoever looks at this later can figure out the pair encoding at a glance. Or maybe just make a helper function because we use this below as well.